File: Algorithms/searching/binary_search.c
/**
* This is an example of linear search and binary search (considering array is sorted)
*/

//Standard includes
#include <stdio.h>
#include <stdbool.h>

/**
 * @brief Perform linear search on a list and find the specified target
 *
 * @param list list of integers
 * @param target target integer to look for
 * @return true if found false otherwise.
 *
 */
bool linearSearch(int list[], int target)
{
    for(int i = 0 ; i < sizeof(list) ; i++)
    {
        if(list[i] == target)
        {
            printf("Found target at index: %d", i);
            return true;
        }
    }
    return false;
}

/**
 * @brief Perform linear search on a list and find the specified target
 *
 * @param list list of integers
 * @param target target integer to look for
 * @return true if found false otherwise.
 *
 */
bool binarySearch(int list[], int target)
{
    for(int i = 0 ; i < sizeof(list) ; i++)
    {
        if(list[i] == target)
        {
            printf("Found target at index: %d", i);
            return true;
        }
    }
    return false;
}

int main(void)
{
    
}

File: Algorithms/searching/binary_search.js
/**
 * @brief Perform binary search on a list and find the specified target
 *
 * @complexity The time complexity is O(log(n)) and the space complexity is O(1)
 *
 * @param list list of integers
 * @param target target integer to look for
 *
 * @return the index if found -1 otherwise.
 *
 */
function binarySearch(list, target) {
  // Assign left and right pointer to be initially beginning of array
  let left = 0;
  let right = list.length - 1;

  // Check if list is not null or undefined
  if (list !== null || list !== undefined) {
    // Loop only while left is smaller or equal than right
    while (left <= right) {
      // Get the middle of the array
      let middle = Math.floor((left + right) / 2);

      // Check if middle is equal to target
      if (list[middle] === target) {
        return middle;
      }

      //Check if the target is smaller than middle
      if (target < list[middle]) right = middle - 1;
      else left = middle + 1;
    }
  }

  return -1;
}

// Test cases
const validArray = [1, 2, 3, 4, 5, 6];
const emptyArray = [];
const stringArray = ["hello", "welcome", "test", "back", "cheetos"];

console.log(binarySearch(emptyArray, 1)); // returns -1
console.log(binarySearch(stringArray, 1)); // returns -1
console.log(binarySearch(validArray, 3)); // returns 2
console.log(binarySearch(validArray, 10)); // returns -1
console.log(binarySearch(validArray, 4)); // returns 3
